From: Martin Schwidefsky <schwidefsky@de.ibm.com>
Subject: kernel: incorrect clock_gettime result
Patch-mainline: v4.0-rc1
Git-commit: 49253925c0be02ed4eb7d94a426731107dd8059d
References: bnc#914291, LTC#121184

Description:  kernel: incorrect clock_gettime result
Symptom:      The value returned by clock_gettime for some clock-ids
              is incorrect.
Problem:      The vdso __kernel_clock_gettime has code to calculate the
              result for CPUCLOCK_VIRT of the currently running thread.
              This is the clock-id -3 but the vdo code incorrectly uses
              the CPUCLOCK_VIRT code for other clock-ids as well.
Solution:     Check for the correct clock-id -3 in the vdso code.
Reproduction: Compare the result of the clock_gettime glibc call with
              the result of a direct system call for clock-id -2.

Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>
Acked-by: John Jolly <jjolly@suse.de>
---
 arch/s390/kernel/vdso64/clock_gettime.S |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

--- a/arch/s390/kernel/vdso64/clock_gettime.S
+++ b/arch/s390/kernel/vdso64/clock_gettime.S
@@ -22,7 +22,7 @@ __kernel_clock_gettime:
 	larl	%r5,_vdso_data
 	cghi	%r2,__CLOCK_REALTIME
 	je	4f
-	cghi	%r2,-2		/* CLOCK_THREAD_CPUTIME_ID for this thread */
+	cghi	%r2,-3		/* Per-thread CPUCLOCK with PID=0, VIRT=1 */
 	je	9f
 	cghi	%r2,__CLOCK_MONOTONIC
 	jne	12f
@@ -81,7 +81,7 @@ __kernel_clock_gettime:
 8:	lghi	%r2,0
 	br	%r14
 
-	/* CLOCK_THREAD_CPUTIME_ID for this thread */
+	/* CPUCLOCK_VIRT for this thread */
 9:	icm	%r0,15,__VDSO_ECTG_OK(%r5)
 	jz	12f
 	ear	%r2,%a4
